h/migrations/versions/8e3417e3713b_back_fill_the_token_user_id_column.py

```diff
@@ -0,0 +1,43 @@
+"""Back-fill the token.user_id column."""
+
+import logging
+import re
+
+from alembic import op
+from sqlalchemy import select
+from sqlalchemy.orm import sessionmaker
+
+from h.models import Token, User
+
+revision = "8e3417e3713b"
+down_revision = "8f5b6f8bbba8"
+
+
+log = logging.getLogger(__name__)
+
+
+def split_userid(userid):
+    return re.match(r"^acct:([^@]+)@(.*)$", userid).groups()
+
+
+def upgrade():
+    session = sessionmaker()(bind=op.get_bind())
+
+    tokens_query = select(Token).where(Token.user_id.is_(None)).limit(1000)
+    count = 0
+
+    while tokens := session.scalars(tokens_query).all():
+        for token in tokens:
+            username, authority = split_userid(token.userid)
+            token.user_id = session.scalars(
+                select(User.id).where(
+                    User.username == username, User.authority == authority
+                )
+            ).one()
+            count += 1
+        session.commit()
+        log.info("Back-filled %d token.user_id's", count)
```

Comment:
There are `312,655` tokens in production right now so this will do `313` separate DB transactions. It'll do `313` `SELECT TOKEN ...` queries, `312,655` `SELECT "user".id ...` queries and `312,655` `UPDATE token ...` queries.

I could add a `time.sleep` in the loop to slow it down, to make sure it doesn't put too much pressure on the DB?

Reply:
For big updates like this without a DB with the same data is difficult to make good guesses.

My intuition it that this will be much faster with less queries and doing more work on SQL instead of python. 

But this seems perfectly doable in the time scale of a migration.

Reply:
It's not speed I'm going for, it's trying to make sure that running the migration doesn't disrupt other DB queries, e.g. by using up too many of the DB's resources at once (e.g. DB CPU usage), or by holding a lock on a table for too long (in this case the `token` table when running the `UPDATE` query). So it's actually the opposite: I'm actually _trying_ to make it go slower, and in separate transactions, so that other requests have a chance to jump in and get their work done in-between these transactions.